perceval/client.py

```diff
@@ -0,0 +1,196 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2015-2017 Bitergia
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
+#
+# Authors:
+#     Valerio Cosentino <valcos@bitergia.com>
+#
+
+import logging
+import time
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+
+from .errors import RateLimitError
+
+logger = logging.getLogger(__name__)
+
+
+class HttpClient:
+    """Abstract class for HTTP clients.
+
+    Base class to query data sources taking care of retrying requests in case connection issues. If the data source does
+    not send back a response after retrying a request, a RetryError exception is thrown.
+
+    Derivated classes can use the methods fetch and _fetch to obtain data from the data source.
+
+    To track which version of the client was used during the fetching
+    process, this class provides a `version` attribute that each client
+    may override.
+
+    :param base_url: base URL of the data source
+    :param max_retries: number of max retries to a data source before raising a RetryError exception
+    :param default_sleep_time: default time to sleep in case of connection problems
+    :param rate_limit_handler: object to handle rate limit issues
+    """
+    version = '0.1'
+
+    DEFAULT_SLEEP_TIME = 1
+    MAX_RETRIES = 5
+
+    MAX_RETRIES_ON_CONNECT = 5
+    MAX_RETRIES_ON_READ = 5
+    MAX_RETRIES_ON_REDIRECT = 5
+    MAX_RETRIES_ON_STATUS = 5
+
+    DEFAULT_RETRY_AFTER_STATUS_CODES = [503, 413, 429]
+
+    GET = "GET"
+    POST = "POST"
+
+    def __init__(self, base_url, max_retries=MAX_RETRIES, default_sleep_time=DEFAULT_SLEEP_TIME):
+        self.base_url = base_url
+        self.session = self.create_http_session()
+        self.max_retries = max_retries
+        self.default_sleep_time = default_sleep_time
+
+    def __del__(self):
+        self.close_http_session()
+
+    def create_http_session(self, headers=None):
+        session = requests.Session()
+
+        if headers:
+            session.headers.update(headers)
+
+        return session
+
+    def close_http_session(self):
+        self.session.close()
+
+    def set_retry_after_status_codes(self, status_codes=None):
```

Comment:
I still don't see why this is needed. I mean, why can't these parameters being set during Retry initialization in line 107?

Reply:
I don't see any parameter to set the retry-after-status-codes during the initialization of the object. 
I have tried to set them up just after the initialization (line 118) but I didn't succeed.

I'm looking at [urllib3](http://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html)